scripts/staging/fm.dml

```diff
@@ -0,0 +1,104 @@
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+# 
+#   http://www.apache.org/licenses/LICENSE-2.0
+# 
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+/*
+ * Factorization Machines.
+ */
+
+forward = function(matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] out) {
+  /*
+   * Computes the model. 
+   *
+   * Reference: 
+   *  - Factorization Machines, Steffen Rendle.
+   *
+   * Inputs:
+   *  - X : n examples with d features, of shape (n, d).
+   *  - w0: the global bias, of shape (1,). <- took this as (n, 1)
+   *  - W : the strength of each feature, of shape (d, 1).
+   *  - V : factorized interaction terms, of shape (d, k).
+   *  
+   * Outputs:
+   *  - out : target vector, of shape (n, 1)
+   */
+   
+   out = w0 + X %*% W + 0.5 * rowSums( (X %*% V)^2 - (X^2 %*% V^2) ); # target vector, shape (n, 1)
+}
+   
+backward = function(matrix[double] dout, matrix[double] X, matrix[double] w0, matrix[double] W, matrix[double] V)
+    return (matrix[double] dX, matrix[double] dw0, matrix[double] dW, matrix[double] dV) {
+   /*
+    * This function accepts the upstream gradients w.r.t output target vector, and 
+    * returns the gradients of the loss w.r.t the parameters
+    *
+    * Inputs:
+    *  - dout : the gradient of the loss function w.r.t y, of shape (n, 1).
+    *  - X, w0, W, V are as mentioned in the above forward function.
+    *
+    * Outputs:
+    *  - dX : the gradient of loss function w.r.t  X, of shape (n, d).
+    *  - dw0: the gradient of loss function w.r.t w0, of shape (1,).
+    *  - dW : the gradient of loss function w.r.t  W, of shape (d, 1). <- make sure about this dim!
+    *  - dV : the gradient of loss function w.r.t  V, of shape (d, k).
+    */
+    
+    # gradient of y (model) w.r.t. X (input)
+    # dX  = ( W + 0.5 * rowSums( (2*X %*% V^2) - (2*X %*% V^2) ) ); # shape (n, d)
+    dX    = ( W + 0.5 * rowsums( (2*X %*% V) %*% V - (2*X %*% V^2) ) ); # needs to be modified 
+ 
+    # gradient w.r.t. w0
+    g_w0 = as.matrix(1);  # shape (1, 1)
+ 
+    dw0  = dout %*% (1 / g_w0) ;  # shape has to be (n, 1)
+    
+    # gradient w.r.t. W
+    g_W = X ; # shape (n, d) 
+    
+    dW  =  (1 / t(g_W)) %*% dout; # shape has to be (d, 1)
+    
+    # gradient w.r.t. V
+    g_V = t(X) %*% rowSums(X %*% V) # - (V %*% X^2) ; # shape (d, 1) - (n, k) :\ not correct!
```

Comment:
@dusenberrymw , second term seems to be a little tricky, where the gradients are not matching.. I tried all possible ways of arranging to get the required dimension, but still something I'm missing out. Please find eq(4) in [paper](https://www.csie.ntu.edu.tw/~b97053/paper/Rendle2010FM.pdf)